service/service.go

```diff
@@ -2,91 +2,161 @@ package service
+	s := serviceList.GetHTTPServer(cfg.BindAddr, r)
+
+	// Get graphDB connection for observation store
+	graphDB, err := serviceList.GetGraphDB(ctx)
+	if err != nil {
+		log.Event(ctx, "failed to initialise graph driver", log.FATAL, log.Error(err))
+		return nil, err
+	}
 
-	s := server.New(cfg.BindAddr, r)
-	s.HandleOSSignals = false
+	// Get dataset API client
+	datasetAPICli := dataset.NewAPIClient(cfg.DatasetAPIURL)
 
-	a := api.Setup(ctx, r)
+	// Setup the API
+	a := api.Setup(ctx, r, cfg, graphDB, datasetAPICli)
 
-	versionInfo, err := healthcheck.NewVersionInfo(
-		buildTime,
-		gitCommit,
-		version,
-	)
+	// Get HealthCheck
+	hc, err := serviceList.GetHealthCheck(cfg, buildTime, gitCommit, version)
 	if err != nil {
-		return nil, errors.Wrap(err, "unable to parse version information")
+		log.Event(ctx, "could not instantiate healthcheck", log.FATAL, log.Error(err))
+		return nil, err
 	}
-	hc := healthcheck.New(versionInfo, cfg.HealthCheckCriticalTimeout, cfg.HealthCheckInterval)
-	if err := registerCheckers(ctx, &hc); err != nil {
+	if err := registerCheckers(ctx, hc, graphDB, datasetAPICli); err != nil {
 		return nil, errors.Wrap(err, "unable to register checkers")
 	}
+
 	r.StrictSlash(true).Path("/health").HandlerFunc(hc.Handler)
 	hc.Start(ctx)
 
+	// Run the http server in a new go-routine
 	go func() {
 		if err := s.ListenAndServe(); err != nil {
 			svcErrors <- errors.Wrap(err, "failure in http listen and serve")
 		}
 	}()
 
 	return &Service{
-		Config:      cfg,
-		Router:      r,
-		API:         a,
-		HealthCheck: &hc,
+		config:      cfg,
+		router:      r,
+		api:         a,
+		healthCheck: hc,
 		server:      s,
+		serviceList: serviceList,
+		graphDB:     graphDB,
 	}, nil
 }
 
-// Gracefully shutdown the service
-func (svc *Service) Close(ctx context.Context) {
-	timeout := svc.Config.GracefulShutdownTimeout
+// Close gracefully shuts the service down in the required order, with timeout
+func (svc *Service) Close(ctx context.Context) error {
+	timeout := svc.config.GracefulShutdownTimeout
 	log.Event(ctx, "commencing graceful shutdown", log.Data{"graceful_shutdown_timeout": timeout}, log.INFO)
 	ctx, cancel := context.WithTimeout(ctx, timeout)
-	defer cancel()
 
-	// stop any incoming requests before closing any outbound connections
-	if err := svc.server.Shutdown(ctx); err != nil {
-		log.Event(ctx, "failed to shutdown http server", log.Error(err), log.ERROR)
-	}
+	// track shutdown gracefully closes app
+	var gracefulShutdown bool
+
+	go func() {
+		defer cancel()
+		var hasShutdownError bool
+
+		// stop healthcheck, as it depends on everything else
+		if svc.serviceList.HealthCheck {
+			svc.healthCheck.Stop()
+		}
+
+		// stop any incoming requests before closing any outbound connections
+		if svc.serviceList.HTTPServer {
+			if err := svc.server.Shutdown(ctx); err != nil {
+				log.Event(ctx, "failed to shutdown http server", log.Error(err), log.ERROR)
+			}
+		}
+
+		// close any API dependency
```

Comment:
Update comment as this is the dataset API

Reply:
yes, at the moment we close all dependencies from the service level (according to the service list boolean values) - this api.Close() was in the skeleton of the project, at the moment it doesn't do anything